public/video-ui/src/slices/targeting.ts

```diff
@@ -1,62 +1,104 @@
+    throw err;
+  })
+);
+
+const debouncedUpdate = debounce(
+  (dispatch, target) =>
+    TargetingApi.updateTarget(target).catch(err => {
+      dispatch(showError(`Failed to update Target`, err));
+      throw err;
+    }),
+  500
+);
+
+export const updateTarget = createAsyncThunk<unknown, TargetType>(
+  'targeting/updateTarget',
+  (target, { dispatch }) => debouncedUpdate(dispatch, target)
+);
+
+export const deleteTarget = createAsyncThunk<unknown, { id: string }>(
+  'targeting/deleteTarget',
+  (target, { dispatch }) =>
+    TargetingApi.deleteTarget(target).catch(err => {
+      dispatch(showError('Failed to delete Target', err));
+      throw err;
+    })
+);
+
+const initialState: TargetingState = { targets: [], deleting: [] };
+
 const targeting = createSlice({
   name: 'targeting',
   initialState,
-  reducers: {
-    requestUpdateTarget:(state:TargetingState, action :PayloadAction<TargetType>)=> ({
-        ...state,
-        targets: [
-          ...(state.targets || []).filter(({ id }) => id !== action.payload.id),
-          action.payload
-        ]
-    }),
-    // when we start a new request for targets reset to our 'loading' state
-    // this gets called when we look for targets for a new video
-    requestGetTargets:()=> ({
-       ...initialState
-    }),
-    // if we receive any targets, for now, completely overwrite what's in here
-    receiveCreateTarget:(state:TargetingState, action :PayloadAction<TargetType>)=> ({
-    ...state,
-        targets: [...(state.targets || []), action.payload]
-    }),
-    receiveGetTarget:(state:TargetingState, action: PayloadAction<TargetType[]>)=> ({
-      ...state,
-      targets: [...(state.targets || []), ...action.payload]
-    }),
-    requestDeleteTarget:(state:TargetingState, action :PayloadAction<TargetType>)=> ({
-      ...state,
-      deleting: [...new Set([...state.deleting, action.payload.id])]
-    }),
-    receiveDeleteTarget:(state:TargetingState, action :PayloadAction<TargetType>)=> ({
-      deleting: [...state.deleting.filter(id => id !== action.payload.id)],
-      targets: [...(state.targets || []).filter(({ id }) => id !== action.payload.id)]
-    }),
-    errorDeleteTarget:(state:TargetingState, action:PayloadAction<TargetType>)=> ({
-      ...state,
-      deleting: [...state.deleting.filter(id => id !== action.payload.id)]
-    })
+  reducers: {},
+  extraReducers: builder => {
+    builder
+      .addCase(getTargets.pending, () => initialState)
+      .addCase(getTargets.fulfilled, (state, { payload }) => {
+        state.targets = [...(state.targets || []), ...payload];
+      })
+      .addCase(createTarget.fulfilled, (state, { payload }) => {
+        state.targets = [...(state.targets || []), payload];
+      })
+      .addCase(updateTarget.pending, (state, { meta }) => {
+        const index = state.targets.findIndex(({ id }) => id === meta.arg.id)
+        index >= 0 && (state.targets[index] = meta.arg);
```

Comment:
Just a question - does the order matter?  The old reducer function appends the target in payload to the end of the `state.targets` (although the new reducer feels more reasonable as the action is to update).